pkg/kubelet/cm/device_plugin_handler.go

```diff
@@ -0,0 +1,257 @@
+	}
+	pdev.podDeviceMapping[podUID][contName].Insert(device)
+}
+
+func (pdev *podDevices) getDevices(podUID, contName string) sets.String {
+	containers, exists := pdev.podDeviceMapping[podUID]
+	if !exists {
+		return nil
+	}
+	devices, exists := containers[contName]
+	if !exists {
+		return nil
+	}
+	return devices
+}
+
+func (pdev *podDevices) delete(pods []string) {
+	for _, uid := range pods {
+		delete(pdev.podDeviceMapping, uid)
+	}
+}
+
+func (pdev *podDevices) devices() sets.String {
+	ret := sets.NewString()
+	for _, containerToDevice := range pdev.podDeviceMapping {
+		for _, deviceSet := range containerToDevice {
+			ret = ret.Union(deviceSet)
+		}
+	}
+	return ret
+}
+
+type DevicePluginHandler interface {
+	Start() error
+	Devices() map[string][]*pluginapi.Device
+	Allocate(pod *v1.Pod, container *v1.Container, activePods []*v1.Pod) ([]*pluginapi.AllocateResponse, error)
+}
+
+// updateCapacityCallback is called to update ContainerManager capacity when
+// device capacity changes.
+type updateCapacityCallback func(v1.ResourceList)
+
+type DevicePluginHandlerImpl struct {
+	sync.Mutex
+	devicePluginManager *deviceplugin.ManagerImpl
+	// allDevices and allocated are keyed by resource_name.
+	allDevices         map[string]sets.String
+	allocatedDevices   map[string]*podDevices
+	updateCapacityFunc updateCapacityCallback
+}
+
+// NewDevicePluginHandler create a DevicePluginHandler
+func NewDevicePluginHandlerImpl(updateCapacityFunc updateCapacityCallback) (*DevicePluginHandlerImpl, error) {
+	glog.V(2).Infof("Starting Device Plugin Handler")
+	handler := &DevicePluginHandlerImpl{
+		allDevices:         make(map[string]sets.String),
+		allocatedDevices:   devicesInUse(),
+		updateCapacityFunc: updateCapacityFunc,
+	}
+
+	deviceManagerMonitorCallback := func(resourceName string, added, updated, deleted []*pluginapi.Device) {
+		var capacity = v1.ResourceList{}
+		kept := append(updated, added...)
+		if _, ok := handler.allDevices[resourceName]; !ok {
+			handler.allDevices[resourceName] = sets.NewString()
+		}
+		// For now, DevicePluginHandler only keeps track of healthy devices.
+		// We can revisit this later when the need comes to track unhealthy devices here.
+		for _, dev := range kept {
+			if dev.Health == pluginapi.Healthy {
+				handler.allDevices[resourceName].Insert(dev.ID)
+			} else {
+				handler.allDevices[resourceName].Delete(dev.ID)
+			}
+		}
+		for _, dev := range deleted {
+			handler.allDevices[resourceName].Delete(dev.ID)
+		}
+		capacity[v1.ResourceName(resourceName)] = *resource.NewQuantity(int64(handler.allDevices[resourceName].Len()), resource.DecimalSI)
+		updateCapacityFunc(capacity)
+	}
+
+	mgr, err := deviceplugin.NewManagerImpl(pluginapi.KubeletSocket, deviceManagerMonitorCallback)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to initialize device plugin manager: %+v", err)
+	}
+
+	handler.devicePluginManager = mgr
+	return handler, nil
+}
+
+func (h *DevicePluginHandlerImpl) Start() error {
```

Comment:
I could not find from where this is being invoked. Should it be invoked from within `NewDevicePluginHandlerImpl()`?

Reply:
No. Object creation is different from logical initialization of an object. The latter may have transitive dependencies on state of other objects and hence requires careful co-ordination initialization.